### PR TITLE
[FSDP2] Added `_to_kwargs` root forward input cast

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118584
* #118298
* #118755
* #118223
* #118136
* #118118
* #118017
* #118004
* #117975
* #117973
* __->__ #117955
* #117950

This PR adds a `_to_kwargs()` call on the FSDP root module's forward inputs to move them to `device` similar to DDP.
https://github.com/pytorch/pytorch/blob/39df084001c54cca5fe3174176f9b0206ddb7dcf/torch/nn/parallel/distributed.py#L1426-L1427

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

### CODE REVIEW

File: torch/distributed/_composable/fsdp/_fsdp_state.py [hunk 1]
```diff
@@ -31,6 +34,13 @@ def _root_pre_forward(
         self._lazy_init()
         if not self._is_root:
             return args, kwargs
+        with torch.profiler.record_function("FSDP::root_pre_forward"):
+            if self._device.type == "cuda":
+                with torch.profiler.record_function("FSDP::inputs_to_device"):
+                    args_tuple, kwargs_tuple = _to_kwargs(
+                        args, kwargs, self._device, False
+                    )  # same as DDP
+                args, kwargs = (args_tuple[0], kwargs_tuple[0])
```
Comment (Skylion007) [inline torch/distributed/_composable/fsdp/_fsdp_state.py:43]:
```suggestion
                args, kwargs = args_tuple[0], kwargs_tuple[0]
```

File: torch/distributed/_composable/fsdp/_fsdp_state.py [hunk 2]
```diff
@@ -71,6 +84,16 @@ def _init_fqns(self) -> None:
             if module in module_to_fsdp_param_group:
                 module_to_fsdp_param_group[module]._module_fqn = module_name
 
+<<<<<<< HEAD
+=======
```
Comment (weifengpy) [inline torch/distributed/_composable/fsdp/_fsdp_state.py:88]: will stamp once resolving this merging conflict

Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_state.py]: Oops 😅  I fixed it now.

**Review by weifengpy — APPROVED** [review]

**Pull request closed without merging**